flang/test/Lower/global-init.f90

```diff
@@ -20,4 +19,14 @@ subroutine foo2()
   my_data = 4
   print *, my_data
 end subroutine
+subroutine foo3()
+! CHECK: fir.address_of(@[[name4:.*foo3.*idata]]){{.*}}fir.array<5xi32>
+! CHECK: fir.address_of(@[[name5:.*foo3.*rdata]]){{.*}}fir.array<3xf16>
+! CHECK: fir.global @[[name4]]{{.*}}fir.array<5xi32>
+! CHECK: fir.global @[[name5]]{{.*}}fir.array<3xf16>
+  integer*4, dimension(5), save :: idata = (/ (i*i, i=1,5) /)
```

Comment:
Is there a semantic check if, for example, i=2,6, if indeed that is incorrect?

Reply:
`(/ (i*i, i=1,5) /)` is equivalent to `(/1,4,9,16,25/)`
So `(/ (i*i, i=2,6) /)` would be correct usage. However,I tried out ``(/ (i*i, i=1,4) /)``, the array literal is one element short of the expected array size and f18 does not generate an error.

```
vjayath @ rome1: build (fir-dev) $ $bld/bin/bbc ../flang/test/Lower/global-init.f90

vjayath @ rome1: build (fir-dev) $ nvfortran ../flang/test/Lower/global-init.f90
NVFORTRAN-S-0066-Too few data constants in initialization statement (../flang/test/Lower/global-init.f90: 27)
  0 inform,   0 warnings,   1 severes, 0 fatal for foo3
```
Of course I tried it with this fir-dev branch, not against f18 in LLVM head.

Reply:
8.6.7p8 says that the lengths of the object list and value list of a DATA statement must be the same.  But it's a program requirement, not a constraint, which might help explain why the front end doesn't currently have a check, if it doesn't.  That's the sort of program requirement that really should be checked if possible.  Meantime, if code to do that checking isn't implemented, and if an eventual front end implementation wouldn't change downstream code such as here, then it might be sufficient for now to put in some variant of an assert.

Reply:
In that case this is not a DATA statement (it's a type-declaration-stmt), but 8.2  point 4 third bullet point also says: "if the entity does not have implied shape, initialization shall either be scalar or have the same shape as the entity.". Given C808 prevents (I think) the shape of the entity from being unknown at compile tome, I am surprised this is not a compiler constraint instead.